peps/pep-0667.rst

```diff
@@ -31,81 +30,268 @@ consistent regardless of threading or coroutines.
+    ...     print(_f_locals["x"])
+    ...
+    >>> f()
+    1
+    1
+    1
+
+This situation was more common in Python 3.10 and earlier
+versions, as merely installing a tracing function was enough
+to trigger implicit calls to ``PyFrame_LocalsToFast()`` after
+every line of Python code. However, it can still happen in Python
+3.11+ depending on exactly which tracing functions are active
+(e.g. interactive debuggers intentionally do this so that changes
+made at the debugging prompt are visible when code execution
+resumes).
+
+All of the above comments in relation to ``exec()`` apply to 
+*any* attempt to mutate the result of ``locals()`` in optimized
+scopes, and are the main reason that the ``locals()`` builtin
+docs contain this caveat:
+
+    Note: The contents of this dictionary should not be modified;
+    changes may not affect the values of local and free variables
+    used by the interpreter. 
+
+Two options were considered to replace this confusing behaviour:
+
+* make ``locals()`` return write-through proxy instances (similar
+  to ``frame.f_locals``)
+* make ``locals()`` return genuinely independent snapshots so that
+  attempts to change the values of local variables via ``exec()``
+  would be *consistently* ignored without any of the caveats
+  noted above.
+
+The PEP chooses the second option for the following reasons:
+
+* returning independent snapshots in optimized scopes preserves
+  the Python 3.0 change to ``exec()`` that resulted in attempts
+  to mutate local variables via ``exec()`` being ignored in most
+  cases
+* the distinction between "``locals()`` gives an instantaneous
+  snapshot of the local variables in optimized scopes, and
+  read/write access in other scopes" and "``frame.f_locals``
+  gives read/write access to the local variables in all scopes,
+  including optimized scopes" allows the intent of a piece of
+  code to be clearer than it would be if both APIs granted
+  full read/write access in optimized scopes, even when write
+  access wasn't needed or desired
+* in addition to improving clarity for human readers, ensuring
+  that name rebinding in optimized scopes remains lexically
+  visible in the code (as long as the frame introspection APIs
+  are not accessed) allows compilers and interpreters to apply
+  related performance optimizations more consistently
+* only Python implementations that support the optional frame
+  introspection APIs will need to provide the new write-through
+  proxy support for optimized frames
+
+This approach is not without its drawbacks, which are covered
+in the Backwards Compatibility section below.
+
 Specification
 =============
 
-Python
-------
+Python API changes
+------------------
+
+The ``frame.f_locals`` attribute
+''''''''''''''''''''''''''''''''
 
 ``frame.f_locals`` will return a view object on the frame that
-implements the ``collections.abc.Mapping`` interface.
+implements the ``collections.abc.MutableMapping`` interface.
```

Comment:
This is incorrect. It should implement the `Mapping` interface. It may be mutable, but it is not a fully mutable mapping. You can't delete some values, and string are supported as keys.

Reply:
Hmm, that's not the current situation either. The current implementation does not allow *any* deletion from the object. It also supports all kinds of keys per your [request](https://github.com/python/cpython/pull/115153#pullrequestreview-2026982160) : ) .

Personally, I believe it's nicer to only support pure strings (we do have an open [issue](https://github.com/python/cpython/issues/120906) about this, allowing other types is kind of a mess for the locals proxy). Originally it was implemented that way because the PEP says so, but if we are editing the PEP, should we reconsider that decision? If we still support all types of keys, we need to answer the issue mentioned - what happens if something that looks like a string got passed in?

Reply:
After checking at the 3.14 interactive prompt, I moved this paragraph to a clearer location and changed it to say:

```
The view objects fully implement the ``collections.abc.Mapping`` interface,
and also implement the following mutable mapping operations:

* using assignment to add new key/value pairs 
* using assignment to update the value associated with a key
* conditional assignment via the `setdefault()` method
* bulk updates via the `update()` method

Removing keys with `del` statements or the `pop()` method is NOT supported.
```

The first three could be simplified to `item assignment (including via the setdefault() method)`, but calling out addition and rebinding separately seemed worthwhile (since they're genuinely different proxy operations)